### PR TITLE
Count number of hops

### PR DESCRIPTION
**Pull request by DavIvek**
### Description
To be able to limit number of hops in query, firstly we need to count number of hops.
Firstly, we want to count number of hops in DFS, BFS and classic expands. Every unit of work is counted as hop, including filtering on edge types, edge properties or visiting already processed nodes.

[master < Task] PR
- [ ] Provide the full content or a guide for the final git message
    - **[FINAL GIT MESSAGE]**

### Documentation checklist
- [ ] Add the documentation label tag
- [ ] Add the bug / feature label tag
- [ ] Add the milestone for which this feature is intended
    - If not known, set for a later milestone
- [ ] Write a release note, including added/changed clauses
    - **[Release note text]**
- [ ] Link the documentation PR here
    - **[Documentation PR link]**
- [ ] Tag someone from docs team in the comments

### CODE REVIEW

File: src/query/interpret/frame.hpp [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2022 Memgraph Ltd.
+// Copyright 2024 Memgraph Ltd.
```
Comment (gitbuda) [inline src/query/interpret/frame.hpp:1]: This should be reverted because nothing else changed in the file 👀

File: src/query/plan/operator.cpp [hunk 2]
```diff
@@ -366,6 +366,7 @@ EdgeAccessor CreateEdge(const EdgeCreationInfo &edge_info, DbAccessor *dba, Vert
                         Frame *frame, ExecutionContext &context, ExpressionEvaluator *evaluator) {
   auto maybe_edge = dba->InsertEdge(from, to, edge_info.edge_type);
   if (maybe_edge.HasValue()) {
+    context.number_of_hops++;
```
Comment (gitbuda) [inline src/query/plan/operator.cpp:369]: I would also skip this for now, only "read" expansions for now 🤔

Comment (gitbuda) [reply src/query/plan/operator.cpp:369]: But whatever we do here, it has to be document in the documentatio.

File: src/query/plan/operator.cpp [hunk 3]
```diff
@@ -1168,14 +1173,18 @@ auto ExpandFromVertex(const VertexAccessor &vertex, EdgeAtom::Direction directio
       memory);
 
   if (direction != EdgeAtom::Direction::OUT) {
-    auto edges = UnwrapEdgesResult(vertex.InEdges(view, edge_types)).edges;
+    auto edges_result = UnwrapEdgesResult(vertex.InEdges(view, edge_types));
+    context->number_of_hops += edges_result.expanded_count;
+    auto edges = std::move(edges_result.edges);
```
Comment (gitbuda) [inline src/query/plan/operator.cpp:1178]: Redundant line, you can just `std::move(edges_result.edges)` two lines further, there are more cases like this in the current version of the code.

**Review by gitbuda — CHANGES_REQUESTED** [review]
@DavIvek A few minor fixes required, but overall looks good (this is the right size of a PR and the right first step 💪)

File: src/query/context.hpp [hunk 4]
```diff
@@ -11,6 +11,7 @@
 
 #pragma once
 
+#include <sys/types.h>
```
Comment (Josipmrden) [inline src/query/context.hpp:14]: do you need this one?

File: src/query/interpreter.cpp [hunk 5]
```diff
@@ -2044,6 +2044,8 @@ std::optional<plan::ProfilingStatsWithTotalTime> PullPlan::Pull(AnyStream *strea
   }
 
   summary->insert_or_assign("plan_execution_time", execution_time_.count());
+  summary->insert_or_assign("number_of_hops", ctx_.number_of_hops);
+  spdlog::trace("Number of hops in the frame: {}", ctx_.number_of_hops);
```
Comment (Josipmrden) [inline src/query/interpreter.cpp:2048]: i think this is a little bit aggressive. maybe to ommit this log

Comment (Josipmrden) [reply src/query/interpreter.cpp:2048]: we already have it in the summary

**Review by Josipmrden — COMMENTED** [review]
some comments but LGTM

**Review by Josipmrden — APPROVED** [review]

**Review by gitbuda — APPROVED** [review]
all comments resolved -> 🚀

**Pull request merged**